megablocks/ops/test_mlp_benchmark.py

```diff
@@ -0,0 +1,205 @@
+
+from absl.testing import parameterized
+from megablocks import benchmark_util
+from megablocks.layers.arguments import Arguments
+from megablocks import ops
+from megablocks.layers.mlp import MLP, SparseMLP, GroupedMLP, TransformerEngineFp8MLP, TorchMLP
+import stk
+
+import torch
+
+def log_benchmark(name, arguments, time, std, flops):
+    benchmark_util.log_benchmark(name, arguments, time, std)
+    print("="*60)
+
+_MATMUL_TESTS = (
+    # (64 * 1024, 512, 2048, 64),
+    # (32 * 1024, 768, 3072, 64),
+    # (8 * 1024, 1024, 4096, 64),
+    # (4 * 2048, 4096, 4 * 4096, 4),
+    (2048, 8192, 8192, 4),
+)
+
+class MLPBenchmark(parameterized.TestCase):
+    def build_input_matrix(self, sl, hs, ne):
+        x = torch.randn((sl, hs), device="cuda", dtype=torch.bfloat16)
+
+        # Assign tokens to experts uniformly.
+        top_expert = torch.arange(0, sl).cuda().int() % ne
+
+        bin_ids, indices = ops.sort(top_expert)
+        tokens_per_expert = ops.histogram(top_expert, ne)
+        padded_tokens_per_expert = ops.round_up(tokens_per_expert, 128)
+        padded_bins = ops.inclusive_cumsum(padded_tokens_per_expert, 0)
+        bins = ops.inclusive_cumsum(tokens_per_expert, 0)
+        out = ops.padded_gather(x, indices, bin_ids, bins, padded_bins, 1)
+        return out, padded_bins
+    
+    def build_sparse_matrix(self, x, padded_bins, fhs, ne):
+        blocking = 128
+        padded_tokens, _ = x.size()
+        assert padded_tokens % blocking == 0
+        assert fhs % blocking == 0
+
+        # Offsets for the sparse matrix. All rows have the
+        # same number of nonzero blocks dictated by the
+        # dimensionality of a single expert.
+        block_rows = padded_tokens // blocking
+        blocks_per_row = fhs // blocking
+        offsets = torch.arange(
+            0,
+            block_rows * blocks_per_row + 1,
+            blocks_per_row,
+            dtype=torch.int32,
+            device=x.device)
+
+        # Indices for the sparse matrix. The indices for
+        # the intermediate matrix are dynamic depending
+        # on the mapping of tokens to experts.
+        column_indices = ops.topology(padded_bins,
+                                      blocking,
+                                      block_rows,
+                                      blocks_per_row)
+        data = torch.empty(
+            column_indices.numel(),
+            blocking,
+            blocking,
+            dtype=torch.float16,
+            device=x.device)
+        shape = (padded_tokens, fhs * ne)
+        row_indices = stk.ops.row_indices(
+            shape, data, offsets, column_indices)
+        return stk.Matrix(shape,
+                          data,
+                          row_indices,
+                          column_indices,
+                          offsets)
+
+
+    def forwardBackward(self, model, x, kwargs=None):
+        y = model(x, **kwargs) if kwargs else model(x)
+        y_grad = y.sum().backward()
```

Comment:
@tgale96 not 100% sure this is the best benchmark to be checking atm. Would love your input.

Afaik, there are three microbenchmarks:
- mlp forward backward
- moe forward backward
- parallel moe forward backward

I'm of the opinion that `parallel moe forward backward` is what we should be microbenchmarking since it'll help us test communication/computation bottlenecks?

Reply:
Mihir and I were saying that you should focus on just the MLP (fwd + bwd). This change only modifies that so we should focus on making that as efficient as possible. Then you can consider end2end model performance separately.

Reply:
Thanks! sounds good